packages/devtools_app/lib/src/shared/preferences/preferences.dart

```diff
@@ -69,44 +99,98 @@ class PreferencesController extends DisposableController
+    final darkModeValue =
+        await storage.getValue(_UiPreferences.darkMode.storageKey);
     final useDarkMode = (darkModeValue == null && useDarkThemeAsDefault) ||
         darkModeValue == 'true';
     ga.impression(gac.devToolsMain, gac.startingTheme(darkMode: useDarkMode));
     toggleDarkModeTheme(useDarkMode);
     addAutoDisposeListener(darkModeEnabled, () {
-      storage.setValue('ui.darkMode', '${darkModeEnabled.value}');
+      storage.setValue(
+        _UiPreferences.darkMode.storageKey,
+        '${darkModeEnabled.value}',
+      );
     });
+  }
 
+  Future<void> _initVmDeveloperMode() async {
     final vmDeveloperModeValue = await boolValueFromStorage(
-      'ui.vmDeveloperMode',
+      _UiPreferences.vmDeveloperMode.storageKey,
       defaultsTo: false,
     );
     toggleVmDeveloperMode(vmDeveloperModeValue);
     addAutoDisposeListener(vmDeveloperModeEnabled, () {
-      storage.setValue('ui.vmDeveloperMode', '${vmDeveloperModeEnabled.value}');
+      storage.setValue(
+        _UiPreferences.vmDeveloperMode.storageKey,
+        '${vmDeveloperModeEnabled.value}',
+      );
     });
+  }
 
-    await _initVerboseLogging();
+  Future<void> _initWasmEnabled() async {
+    wasmEnabled.value = kIsWasm;
+
+    // It is important that this listener is added before we set the initial
+    // state of the wasm mode setting below. This is because the query parameter
+    // for wasm may need to be updated based on the value of the preference in
+    // the storage file, which we take into account when we call
+    // [toggleWasmEnabled] at the end of this method.
+    addAutoDisposeListener(wasmEnabled, () async {
+      final enabled = wasmEnabled.value;
+      await storage.setValue(
+        _ExperimentPreferences.wasm.storageKey,
+        '$enabled',
+      );
 
-    await inspector.init();
-    await memory.init();
-    await logging.init();
-    await performance.init();
-    await devToolsExtensions.init();
+      // Update the wasm mode query parameter if it does not match the value of
+      // the setting.
+      final wasmEnabledFromQueryParams = DevToolsQueryParams.load().useWasm;
+      if (wasmEnabledFromQueryParams != enabled) {
+        updateQueryParameter(
+          DevToolsQueryParams.wasmKey,
+          enabled ? 'true' : null,
+          reload: true,
+        );
+      }
+    });
 
-    setGlobal(PreferencesController, this);
+    final enabledFromStorage = await boolValueFromStorage(
+      _ExperimentPreferences.wasm.storageKey,
+      defaultsTo: false,
+    );
+    final enabledFromQueryParams = DevToolsQueryParams.load().useWasm;
+
+    // TODO(kenz): this may cause an infinite loop of reloading the page if
+    // the setting from storage or the query parameter indicate we should be
+    // loading with WASM, but each time we reload the page, something goes wrong
+    // and we fall back to JS.
+    toggleWasmEnabled(enabledFromStorage || enabledFromQueryParams);
```

Comment:
@eyebrowsoffire @yjbanov is there a way to know that we tried to use WASM but had to fall back? Otherwise this could get in an infinite loop when the attempt to load with WASM fails and we fell back to JS. 

This would make `kIsWasm` false, and we would continually try to change the value of the `wasmEnabled` notifier to true based on the value of `(enabledFromStorage || enabledFromQueryParams)`. This would then trigger the listener and reload the page again, and repeat.

Reply:
I think you can just check the actual query params via JS interop (I'm not sure if `DevToolsQueryParams` is actually doing that or if it is some other system for composing the URL to pass back to the user) and then also look at `kIsWasm`. If the query parameters in the actual url have the `useWasm` parameter, but you're not in wasm mode, then we've fallen back to JS.

I have contemplated exposing some more rich information about what happened in the bootstrap process and why, but I don't know exactly what that would look like and we haven't implemented anything currently.